File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int result = n;
        
        for (int i = 0; i < wires.length; i++) {
            Set<Integer> nodes = new HashSet<>();
            if (i != 0) {
                nodes.add(wires[0][0]);
                nodes.add(wires[0][1]);
            } else {
                nodes.add(wires[1][0]);
                nodes.add(wires[1][1]);
            }

            for (int j = 0; j < wires.length; j++) {
                for (int k = 0; k < wires.length; k++) {
                    if (i != k) {
                        if (nodes.contains(wires[k][0]) || nodes.contains(wires[k][1])) {
                            nodes.add(wires[k][0]);
                            nodes.add(wires[k][1]);
                        }
                    }
                }
            }
            result = (int) Math.min(result, Math.abs(n - 2 * nodes.size()));
        }
        return result;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict

def solution(n, wires):

    graph = defaultdict(list)
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)

    visited = [False] * (n+1)
    children = [0] * (n + 1)
    
    find_children(1, children, visited, graph)

    return min(abs(n - 2 *(children[i] + 1)) for i in range(1, n+1))

    
def find_children(node, children, visited, graph):
    visited[node] = True
    result = 0
    for next_node in graph[node]:
        if not visited[next_node]:
            result += 1 + find_children(next_node, children, visited, graph)

    children[node] = result
    return result